### DISCUSSION TITLE
Customizable log_format string

### DISCUSSION THREAD

**Discussion by JMLX42** [nginx-proxy/nginx-proxy#2133] [Ideas]
## Rationale

Today, the `log_format` is hardcoded:

https://github.com/nginx-proxy/nginx-proxy/blob/2f898566fcf20f36e29a6265dc42e56587c59eba/nginx.tmpl

Some might need to use a different format string. In our case, we need to include `$request_time` because we want to use Loki + Graphana to chart our APIs response time.

AFAIK:

- there is no way to customize the `log_format` string without editing the template
- there is no (documented) way to extend the template without (as opposed to completely override it) to change the value of `$access_log`

## Proposal

Add the `VIRTUAL_LOG_FORMAT` and `DEFAULT_LOG_FORMAT` environment variables. Such environment variable would be set to the `string` expected argument of the `log_format` NGINX directive:

https://nginx.org/en/docs/http/ngx_http_log_module.html#log_format

```
log_format name [escape=default\|json\|none] string ...
```

Example to add `$request_time` to the access logs:

```
DEFAULT_LOG_FORMAT='$host $remote_addr - $remote_user [$time_local] "$request" $status $body_bytes_sent $request_time "$http_referer" "$http_user_agent" "$upstream_addr"'
```

- If `VIRTUAL_LOG_FORMAT` use it. Otherwise use `DEFAULT_LOG_FORMAT`.
- If `DEFAULT_LOG_FORMAT` is not set, use the hardcoded default value.

Thus, this addition would be 100% backward compatible.

**Comment by vincent-herlemont** [comment]
@JMLX42 Here is a first simplified implementation https://github.com/nginx-proxy/nginx-proxy/pull/2151.

**Reply by buchdag** [reply]
@JMLX42 done in #2151 thanks to @vincent-herlemont

**Reply by JMLX42** [reply]
@buchdag yes awesome! Can we expect a minor release anytime soon?

**Reply by buchdag** [reply]
I'd say by the end of the week, we're just waiting for #1179 and it's almost done thanks to @rhansen

**Comment by rhansen** [comment]
> Add the `VIRTUAL_LOG_FORMAT` and `DEFAULT_LOG_FORMAT` environment variables.

@JMLX42 Why two environment variables?  #2151 only uses one, and it seems like a suitable approach to me.

**Reply by JMLX42** [reply]
> Why two environment variables? https://github.com/nginx-proxy/nginx-proxy/pull/2151 only uses one, and it seems like a suitable approach to me.

@rhansen one might want to both:

1. Override the template's log format for all virtual host.
2. Override the default log format for a specific virtual host.

It's fine to use the same variable name for both 1. and 2.

But unless I am mistaken, use case 2. is not yet covered by #2151.

PS: disclaimer @vincent-herlemont is a colleague of mine and we're both working on this for the same project
PS 2: in our case, we do *not* need a per-virtual host log format. So I won't fight that much for it really :smile:

**Reply by rhansen** [reply]
For per-vhost config I'd prefer the same variable name, just applied to the upstream application container instead of the `nginx-proxy` container.   (Actually, it would probably be a label, not an environment variable, on the upstream container; see #2148.)

**Comment by rhansen** [comment]
I haven't tested it, but I think you can currently customize the log format by doing the following:

  * Set `DISABLE_ACCESS_LOGS=true` environment variable on the `nginx-proxy` container.
  * Mount the following to `/etc/nginx/proxy.conf`:
    ```
    log_format custom 'foo bar baz';
    access_log /var/log/nginx/access.log custom
    ```

**Reply by JMLX42** [reply]
> I haven't tested it, but I think you can currently customize the log format by doing the following:

@rhansen most likely correct (cf https://github.com/nginx-proxy/nginx-proxy#replacing-default-proxy-settings). But it implies:

> NOTE: If you provide this file it will replace the defaults; you may want to check the .tmpl file to make sure you have all of the needed options.

Which is troublesome just to change something as common/trivial as configuring the log format.

**Comment by darh** [comment]
I've looked around on how JSON-formatted logging can be configured. Even the latest changes/commits do not include that `escape` param for `log_format` directive. Maybe I'm missing something.

I've tinkered around a bit with the log formatting configuration.

https://github.com/darh/nginx-proxy/commit/cc767e1710dfa5ff1f1b1fc0bde243aaceb857f0

I can polish it further if needed. Let me know.

**Reply by rhansen** [reply]
You're not missing anything @darh; the lack of support for `escape=*` is an oversight.  Sigh.

Rather than JSON-specific settings, what do you think about a new `LOG_FORMAT_ESCAPE` setting that the user can set to `json` if desired?  That would accommodate `escape=none` as well as any escaping schemes nginx might add in the future.

The `LOG_JSON=true` shorthand is nice.  Maybe we could change the default value of `LOG_FORMAT` to depend on the value of `LOG_FORMAT_ESCAPE`.  Or is that too magical?

If backwards compatibility wasn't a concern, I'd be tempted to remove the single quotes that currently exist in the template and force the user to provide any necessary quoting/escaping.  Then it would be possible to do something like `LOG_FORMAT="escape=json 'foo bar baz'"`.

**Reply by rhansen** [reply]
Another thought: skip the single quotes if `LOG_FORMAT` ends with a single- or double-quote character.  Then people could do `LOG_FORMAT="escape=json 'foo var baz'"` without breaking most (all?) users.  (Theoretically, users might currently end their `LOG_FORMAT` with `"` or `\'`.  Not sure how likely that is.)

**Reply by darh** [reply]
I've pushed an update. See commit message for details https://github.com/darh/nginx-proxy/commit/56c7a3bbb7644245c083897d1396eadf18621c7a

Log formatting can now be done through `LOG_FORMAT`, `LOG_FORMAT_ESCAPE` and  `LOG_JSON`.

**Reply by azN2** [reply]
thanks for this, i needed json log format

**Comment by darh** [comment]
I'm sure people might not like us removing ' and breaking old configs. 

Happy to redo that PR to `LOG_FORMAT_ESCAPE` and make LOG_FORMAT default dependable on it.

**Reply by buchdag** [reply]
I confirm that breaking old configs must be avoided at this point. 👍
